File: app/src/main/java/org/thoughtcrime/securesms/blocked/BlockedUsersViewModel.java
package org.thoughtcrime.securesms.blocked;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import org.thoughtcrime.securesms.recipients.Recipient;
import org.thoughtcrime.securesms.recipients.RecipientId;

import java.util.List;
import java.util.Objects;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.PublishSubject;
import io.reactivex.rxjava3.subjects.Subject;

public class BlockedUsersViewModel extends ViewModel {

  private final BlockedUsersRepository   repository;
  private final Subject<List<Recipient>> recipients = BehaviorSubject.create();
  private final Subject<Event>           events     = PublishSubject.create();

  private BlockedUsersViewModel(@NonNull BlockedUsersRepository repository) {
    this.repository  = repository;

    loadRecipients();
   }

  public Observable<List<Recipient>> getRecipients() {
    return recipients.observeOn(AndroidSchedulers.mainThread());
  }

  public Observable<Event> getEvents() {
    return events.observeOn(AndroidSchedulers.mainThread());
  }

  void block(@NonNull RecipientId recipientId) {
    repository.block(recipientId,
                     () -> {
                       loadRecipients();
                       events.onNext(new Event(EventType.BLOCK_SUCCEEDED, Recipient.resolved(recipientId)));
                     },
                     () -> events.onNext(new Event(EventType.BLOCK_FAILED, Recipient.resolved(recipientId))));
  }

  void createAndBlock(@NonNull String number) {
    repository.createAndBlock(number, () -> {
      loadRecipients();
      events.onNext(new Event(EventType.BLOCK_SUCCEEDED, number));
    });
  }

  void unblock(@NonNull RecipientId recipientId) {
    repository.unblock(recipientId, () -> {
      loadRecipients();
      events.onNext(new Event(EventType.UNBLOCK_SUCCEEDED, Recipient.resolved(recipientId)));
    });
  }

  private void loadRecipients() {
    repository.getBlocked(recipients::onNext);
  }

  enum EventType {
    BLOCK_SUCCEEDED,
    BLOCK_FAILED,
    UNBLOCK_SUCCEEDED
  }

  public static final class Event {

    private final EventType eventType;
    private final Recipient recipient;
    private final String    number;

    private Event(@NonNull EventType eventType, @NonNull Recipient recipient) {
      this.eventType = eventType;
      this.recipient = recipient;
      this.number    = null;
    }

    private Event(@NonNull EventType eventType, @NonNull String number) {
      this.eventType = eventType;
      this.recipient = null;
      this.number    = number;
    }

    public @Nullable Recipient getRecipient() {
      return recipient;
    }

    public @Nullable String getNumber() {
      return number;
    }

    public @NonNull EventType getEventType() {
      return eventType;
    }
  }

  public static class Factory implements ViewModelProvider.Factory {

    private final BlockedUsersRepository repository;

    public Factory(@NonNull BlockedUsersRepository repository) {
      this.repository = repository;
    }

    @Override
    public @NonNull <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
      return Objects.requireNonNull(modelClass.cast(new BlockedUsersViewModel(repository)));
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/blocked/BlockedUsersViewModel.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.thoughtcrime.securesms.blocked

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.thoughtcrime.securesms.recipients.Recipient
import org.thoughtcrime.securesms.recipients.Recipient.Companion.resolved
import org.thoughtcrime.securesms.recipients.RecipientId


class BlockedUsersViewModel(private val repository: BlockedUsersRepository) : ViewModel() {
  private val _events = MutableSharedFlow<BlockUserEvent>()
  val events = _events.asSharedFlow()

  private val _blockedUsers = MutableStateFlow<List<BlockedUserRecipientState>>(emptyList())
  val blockedUsers = _blockedUsers.asStateFlow()

  init {
    viewModelScope.launch {
      loadRecipients()
    }
  }

  fun block(recipientId: RecipientId) {
    viewModelScope.launch {
      repository.block(recipientId)
        .onSuccess {
          loadRecipients()
          _events.emit(BlockUserEvent.BlockSucceeded(resolved(recipientId)))
        }.onFailure {
          _events.emit(BlockUserEvent.BlockFailed(resolved(recipientId)))
        }
    }
  }

  fun createAndBlock(number: String) {
    viewModelScope.launch {
      repository.createAndBlock(number)
        .onSuccess {
          loadRecipients()
          _events.emit(BlockUserEvent.CreateAndBlockSucceeded(number))
        }
    }
  }

  fun unblock(recipientId: RecipientId) {
    viewModelScope.launch {
      repository.unblock(recipientId)
        .onSuccess {
          loadRecipients()
          _events.emit(BlockUserEvent.UnblockSucceeded(resolved(recipientId)))
        }
    }
  }

   private suspend fun loadRecipients() {
      repository.getBlocked()
        .onSuccess { recipientList ->
          _blockedUsers.update { recipientList }
      }
  }

}

sealed interface BlockUserEvent {
  data class BlockSucceeded(val recipient: Recipient) : BlockUserEvent
  data class BlockFailed(val recipient: Recipient): BlockUserEvent
  data class CreateAndBlockSucceeded(val number : String) : BlockUserEvent
  data class UnblockSucceeded(val recipient: Recipient): BlockUserEvent
}